### PR TITLE
Build: Add font generation to the grunt process

### PR DESCRIPTION
**Pull request by jasmussen**
This automates the build process. To generate a font, now make sure you've run `npm update`, then run `npm run build`.

The `codepoints.json` (props @field2) should ensure that existing icons are mapped to the same unicode locations as the old font was. But we need to verify this thoroughly.

Now that the font generation is automated, we should consider retiring/deleting these folders:

- `source`
- `js`
- `fonts`
- `css`

At a glance it looks like all of those are obsolete now. Is that correct?

### CODE REVIEW

**Comment by jasmussen** [comment]
By "thorough review", we need to try and use the new icon font in place of the old one, and see that everything still looks right. It _should_ be a drop-in replacement. Perhaps a good way to test is to look at the old example page, replace the fonts there, and then see that every icon still looks right.

**Comment by jasmussen** [comment]
In order to expediate getting this merged, I tried to compare the two fonts, and use the grunt built one as a drop in replacement on the dashicons dev page. Here are screenshots you can compare:

Glyphs, baseline:

glyphs-dashicons

Grunt version:

grunt-dashicons

Here is a MASSIVE GIF file which compares the two overlaid on top of each other, switching between the two:

comparison

This quickly makes a few things clear:

- The grunt version has a lower baseline, so icons sit a bit lower than the other. That's fine, I think . 
- The `menu` icon is different in the new version. Did the SVG file change, but the Glyphs file wasn't updated?
- There's something wrong with the fullscreen icon in the grunt version.

@field2 can you take a look?

More GIFS:

menu-comparison

fullscreen

**Comment by jasmussen** [comment]
As a followup, it looks like there's no `expand.svg` file in the SVG folder at all. That would explain why it's missing :)

**Comment by field2** [comment]
In the svg's, there are 2 files: menu.svg and menu-alt.svg. Menu-alt.svg is closer to the one in the Glyphs file, so I think we should just rename/switch their names.

**Comment by ryelle** [comment]
I agree that the listed folders are obsolete now — at least, I think css & js have been obsolete for a while.

While running this I saw:
```
The glyph named plus is mapped to U+F132.
But its name indicates it should be mapped to U+002B.

The glyph named minus is mapped to U+F460.
But its name indicates it should be mapped to U+2212.
The glyph named heart is mapped to U+F487.
But its name indicates it should be mapped to U+2665.
```
I assume this is an automated check that we have icon names matching existing unicode chars, and it's suggesting we map to those (and we don't want to do that). Is that correct?

Are there Windows install/build instructions? The brew commands are mac-specific, do windows users not need ttfautohint & fontforge?

Other than those questions, this looks good from a code perspective 🙂  (I also left some comments on #169 about the build process, but those _shouldn't_ affect anything here)

**Comment by jasmussen** [comment]
> I agree that the listed folders are obsolete now — at least, I think css & js have been obsolete for a while.

👍 

> I assume this is an automated check that we have icon names matching existing unicode chars, and it's suggesting we map to those (and we don't want to do that). Is that correct?

That seems correct, yes, as "plus" is <a href="http://www.fileformat.info/info/unicode/char/2b/index.htm">U+002B</a>. However back in the day I would map genericons to the closest looking unicode match, and this turned out to be a bad idea for a number of reasons, including a "brief flash of unstyled code". Since then I think icon fonts have largely used the extended unicode space, where unstyled text would show empty glyphs.

> Are there Windows install/build instructions? The brew commands are mac-specific, do windows users not need ttfautohint & fontforge?

We should probably add that, yeah. We can copy some from <a href="https://github.com/sapegin/grunt-webfont">here</a>.

File: codepoints.json [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    "thumbs-up": 62761,
+    "thumbs-down": 62786,
+    "layout": 62776,
+    "paperclip": 62790,
+    "email-alt2": 61706,
+    "menu-alt": 61707,
+    "plus-light": 61708,
+    "trash": 61709,
+    "uniF143": 61710,
+    "uniF494": 61711
```
Comment (ryelle) [inline codepoints.json]: These should probably get the correct names (or removed, if they're duplicates):

screen shot 2017-04-12 at 2 53 10 pm

Comment (jasmussen) [inline codepoints.json]: CC: @field2 do these ring bells?

**Comment by ryelle** [comment]
Does this auto-generate the codepoints file? After running build on this PR, I now have `"editor-expand": 61715` added to codepoints.json, which I thought had to be manually created.

**Comment by jasmussen** [comment]
> Does this auto-generate the codepoints file? After running build on this PR, I now have "editor-expand": 61715 added to codepoints.json, which I thought had to be manually created.

Any new icons are automatically appended to the file, so their places will be reserved. That is, if you later remove an icon, then add a new one, the old codepoint won't be used, causing incompatibilities in some bad CSS implementation that specified a specific one. 

This is also what allowed us to create 1:1 drop-in back compatibility with the old dashicons font, so icons keep the same unicode locations as they did there.

File: codepoints.json (hunk 1 above)
Comment (jasmussen) [inline codepoints.json]: Addressed this. They were duplicates. Solid catch!

**Comment by jasmussen** [comment]
Addressed the feedback:

- Rebased
- Removed duplicate icons
- Removed obsolete folders
- Added Windows & Linux instructions

Good to go?

**Pull request merged**